core/preconditioner/batch_ilu_isai.cpp

```diff
@@ -0,0 +1,176 @@
+            .on(exec);
+
+    // Note: Use a custom deleter to prevent memory deallocation when the
+    // control returns to calling function and smart pointer goes out of scope
+    // When this smart pointer is assigned to another shared ptr, its deleter is
+    // also copied.
+    std::shared_ptr<const BatchLinOp> sys_smart_ptr(
+        system_matrix, [](const BatchLinOp* plain_ptr) {});
+
+    auto batch_ilu_precond = batch_ilu_factory->generate(sys_smart_ptr);
+
+    std::pair<std::shared_ptr<const matrix_type>,
+              std::shared_ptr<const matrix_type>>
+        l_and_u_factors =
+            batch_ilu_precond->generate_split_factors_from_factored_matrix();
+
+    std::shared_ptr<const matrix_type> l_factor = l_and_u_factors.first;
+    std::shared_ptr<const matrix_type> u_factor = l_and_u_factors.second;
+    this->lower_factor_ = l_factor;
+    this->upper_factor_ = u_factor;
+
+    // Note: The generated l_factor and u_factor are always sorted irrespective
+    // of whether the input matrix is sorted or unsorted.
+    auto lower_factor_isai_precond =
+        gko::preconditioner::BatchIsai<ValueType, IndexType>::build()
+            .with_skip_sorting(true)
+            .with_isai_input_matrix_type(
+                batch_isai_input_matrix_type::lower_tri)
+            .with_sparsity_power(
+                this->parameters_.lower_factor_isai_sparsity_power)
+            .on(exec)
+            ->generate(l_factor);
+
+    auto upper_factor_isai_precond =
+        gko::preconditioner::BatchIsai<ValueType, IndexType>::build()
+            .with_skip_sorting(true)
+            .with_isai_input_matrix_type(
+                batch_isai_input_matrix_type::upper_tri)
+            .with_sparsity_power(
+                this->parameters_.upper_factor_isai_sparsity_power)
+            .on(exec)
+            ->generate(u_factor);
+
+    this->lower_factor_isai_ =
+        lower_factor_isai_precond->get_const_approximate_inverse();
+    this->upper_factor_isai_ =
+        upper_factor_isai_precond->get_const_approximate_inverse();
+
+    // Note: mult_inv and iteration matrices etc. are created/computed outside
+    // the solver kernel (in precond. external generate)
+    // since they require SpGemm (memory allocation to store the result is not
+    // straightforward)
+
+    if (this->parameters_.apply_type ==
+        batch_ilu_isai_apply::spmv_isai_with_spgemm) {
+        // z = precond * r
+        // L * U * z = r
+        // lai_L * L * U * z = lai_L * r
+        // U * z = lai_L * r
+        // lai_U * U * z = lai_U * lai_L * r
+        // z = lai_U * lai_L * r
+        // z = mult_inv * r
+        // Therefore, mult_inv = lai_U * lai_L
+        this->mult_inv_ =
+            gko::share(matrix_type::create(exec, system_matrix->get_size()));
+        this->upper_factor_isai_->apply(this->lower_factor_isai_.get(),
+                                        this->mult_inv_.get());
```

Comment:
The `batch_ilu_isai_apply::spmv_isai_simple` approach does exactly that. I just thought of having this SpGeMM approach (`batch_ilu_isai_apply::spmv_isai_with_spgemm`) to avoid two SpMVs in the preconditioner apply step. SpGEMM computation is expensive but  it is computed only once and it allows to do preconditioner application, which would be done multiple times in the iterative solver, with just one SpMV each time instead of two.

Reply:
I did not really check as the SpGEMM implementation is not yet there. But maybe I leave it here for the time being.

Reply:
Yes, I think that is fine. We should also maybe think about whether an SPD ISAI makes sense for batched.